src/photon/photon_algorithm.c

```diff
@@ -168,78 +182,144 @@ void fetch_missing_dependencies(local_scheduler_state *state,
 }
 
 /**
- * If there is a task whose dependencies are available locally, assign it to the
- * worker. This does not remove the worker from the available worker queue.
+ * Assign as many tasks from the dispatch queue as possible.
  *
- * @param s The scheduler state.
- * @param worker_index The index of the worker.
- * @return This returns 1 if it successfully assigned a task to the worker,
- *         otherwise it returns 0.
+ * @param state The scheduler state.
+ * @param algorithm_state The scheduling algorithm state.
+ * @return Void.
  */
-bool find_and_schedule_task_if_possible(
-    local_scheduler_state *state,
-    scheduling_algorithm_state *algorithm_state,
-    int worker_index) {
-  task_queue_entry *elt, *tmp;
-  bool found_task_to_schedule = false;
-  /* Find the first task whose dependencies are available locally. */
-  DL_FOREACH_SAFE(algorithm_state->task_queue, elt, tmp) {
-    if (can_run(algorithm_state, elt->spec)) {
-      found_task_to_schedule = true;
-      break;
-    }
-  }
-  if (found_task_to_schedule) {
-    /* This task's dependencies are available locally, so assign the task to the
-     * worker. */
-    assign_task_to_worker(state, elt->spec, worker_index,
-                          elt->from_global_scheduler);
-    /* Update the task queue data structure and free the task. */
-    DL_DELETE(algorithm_state->task_queue, elt);
-    free_task_spec(elt->spec);
-    free(elt);
-  }
-  return found_task_to_schedule;
-}
+void dispatch_tasks(local_scheduler_state *state,
+                    scheduling_algorithm_state *algorithm_state) {
+  /* Assign tasks while there are still tasks in the dispatch queue and
+   * available workers. */
+  while ((algorithm_state->dispatch_task_queue != NULL) &&
+         (utarray_len(algorithm_state->available_workers) > 0)) {
+    LOG_DEBUG("Dispatching task");
+    /* Pop a task from the dispatch queue. */
+    task_queue_entry *dispatched_task = algorithm_state->dispatch_task_queue;
```

Comment:
Yes. According to the utlist documentation, the pointer points to the head, or the first element.